AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Subscribe/DataStoreObserveQueryOperation.swift

```diff
@@ -318,21 +311,27 @@ public class AWSDataStoreObserveQueryOperation<M: Model>: AsynchronousOperation,
 
     func generateQuerySnapshot(itemsChanged: [MutationEvent] = []) {
         updateCurrentItems(with: itemsChanged)
-        var currentItems = Array(currentItemsMap.values.map { $0 }) as [M]
         if let sort = sortInput {
             sort.forEach { currentItems.sortModels(by: $0, modelSchema: modelSchema) }
         }
-        publishSnapshot(ofItems: currentItems, isSynced: isSynced.get(), itemsChanged: itemsChanged)
+        passthroughPublisher.send(currentSnapshot)
     }
 
     func updateCurrentItems(with itemsChanged: [MutationEvent]) {
         for item in itemsChanged {
             do {
                 let model = try item.decodeModel(as: modelType)
                 if item.mutationType == MutationEvent.MutationType.delete.rawValue {
-                    currentItemsMap.removeValue(forKey: model.id)
+                    if let index = currentItems.firstIndex(where: { $0.id == model.id }) {
+                        currentItems.remove(at: index)
```

Comment:
i think `currentItems` should be thread safe to avoid duplicates. and a unit test to check for duplicates case.

Reply:
It may not be obvious since this method itself is not thread safe but it's run under the serial DispatchQueue which will perform each task under `serialQueue.async` serially, so it doesn't have to be thread safe. As long as the entry points put the work on the serialQueue, then the work will run one after another and no two threads will be accessing `currentItems` at the same time. There's a `testConcurrent` test DataStoreObserveQueryOperationTests which tests this